BEIMA.Backend/DeviceFunctions/UpdateDevice.cs

```diff
@@ -38,33 +55,85 @@ public static class UpdateDevice
                 return new BadRequestObjectResult(Resources.InvalidIdMessage);
             }
 
+            var reqForm = await req.ReadFormAsync();
+            var mongo = MongoDefinition.MongoInstance;
+
+            // Parse out necessary data
+            UpdateDeviceRequest data;
             Device device;
+            ObjectId deviceTypeId;
+            ObjectId buildingId;
             try
             {
-                string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
-                dynamic data = JsonConvert.DeserializeObject(requestBody);
-                device = new Device(new ObjectId(id),
-                                    ObjectId.Parse((string)data.deviceTypeId),
-                                    (string)data.deviceTag,
-                                    (string)data.manufacturer,
-                                    (string)data.modelNum,
-                                    (string)data.serialNum,
-                                    (int)data.yearManufactured,
-                                    (string)data.notes);
-
-                device.SetLocation(ObjectId.Parse((string)data.location.buildingId),
-                                   (string)data.location.notes,
-                                   (string)data.location.latitude,
-                                   (string)data.location.longitude);
-
-                device.SetLastModified(DateTime.UtcNow, "Anonymous");
-                // TODO: include device type attributes
+                data = JsonConvert.DeserializeObject<UpdateDeviceRequest>(reqForm["data"]);
+                deviceTypeId = ObjectId.Parse(data.DeviceTypeId);
+                buildingId = ObjectId.Parse(data.Location.BuildingId);
+
+                var deviceDocument = mongo.GetDevice(new ObjectId(id));
+                device = BsonSerializer.Deserialize<Device>(deviceDocument);
             }
             catch (Exception)
             {
                 return new BadRequestObjectResult(Resources.CouldNotParseBody);
             }
 
+            // Set Data to new values
+            device.DeviceTypeId = deviceTypeId;
+            device.DeviceTag = data.DeviceTag;
+            device.Manufacturer = data.Manufacturer;
+            device.ModelNum = data.ModelNum;
+            device.SerialNum = data.SerialNum;
+            device.YearManufactured = data.YearManufactured;
+            device.Notes = data.Notes;
+            device.SetLocation(
+                buildingId,
+                data.Location.Notes,
+                data.Location.Latitude,
+                data.Location.Longitude
+            );
+            device.Fields = data.Fields;
+
+            // Check if there is a new photo, if so remove all others on the device
+            // Frontend only supports 1 photo, but a device can contain more
+            var updatePhoto = reqForm.Files.Any(file => file.Name == "photos");
+            if (updatePhoto && device.Photos.Count > 0)
```

Comment:
It feels weird to me to just delete all photos if there is any photos in the request. It seems like we should be able to individually select which photos to remove, add, or keep.

Reply:
Discussed outside of this PR, for now we will only allow users to upload a single photo rather than a set of photos.